### PR TITLE
Attempt to fix Health Connect on Android 14

### PR DESCRIPTION
**Pull request by eliasteeny**
This pull request addresses the "Activity not found" error encountered when attempting to open the Health Connect permissions screen on Android 14. The issue was resolved by migrating from the old `startActivityForResult` method to the new API, `registerForActivityResult`.

Introduced a limitation for the Main app's activity: it should extend `FlutterFragmentActivity` instead of `FlutterActivity`. This adjustment allows casting from `Activity` to `ComponentActivity` for accessing `registerForActivityResult`.

Note: As a non-experienced Android/Kotlin developer, any suggestions for improvements or alternative approaches are welcome especially regarding `FlutterFragmentActivity`.

### CODE REVIEW

**Comment by chadpav** [comment]
I was able to pull down this PR branch and verify the functionality work. The code looks reasonable to me.

This PR resolves Issue #800 

Thank you @eliasteeny

File: packages/health/example/lib/main.dart [hunk 1]
```diff
@@ -160,14 +160,14 @@ class _HealthAppState extends State<HealthApp> {
...
     success &= await health.writeBloodPressure(90, 80, earlier, now);
-    success &= await health.writeHealthData(
-        0.0, HealthDataType.SLEEP_REM, earlier, now);
-    success &= await health.writeHealthData(
-        0.0, HealthDataType.SLEEP_ASLEEP, earlier, now);
-    success &= await health.writeHealthData(
-        0.0, HealthDataType.SLEEP_AWAKE, earlier, now);
-    success &= await health.writeHealthData(
-        0.0, HealthDataType.SLEEP_DEEP, earlier, now);
+    // success &= await health.writeHealthData(
```
Comment (chadpav) [inline packages/health/example/lib/main.dart:163]: Are these changes relevant to the PR?

File: packages/health/example/lib/util.dart [hunk 2]
```diff
@@ -68,12 +68,12 @@ const List<HealthDataType> dataTypesAndroid = [
   // HealthDataType.MOVE_MINUTES, // TODO: Find alternative for Health Connect
   HealthDataType.DISTANCE_DELTA,
   HealthDataType.RESPIRATORY_RATE,
-  HealthDataType.SLEEP_AWAKE,
-  HealthDataType.SLEEP_ASLEEP,
-  HealthDataType.SLEEP_LIGHT,
-  HealthDataType.SLEEP_DEEP,
-  HealthDataType.SLEEP_REM,
-  HealthDataType.SLEEP_SESSION,
+  // HealthDataType.SLEEP_AWAKE,
```
Comment (chadpav) [inline packages/health/example/lib/util.dart:71]: Are these changes relevant?

File: packages/health/example/android/app/src/main/kotlin/cachet/plugins/example_app/MainActivity.kt [hunk 3]
```diff
@@ -2,7 +2,7 @@ package cachet.plugins.example_app
 
 import android.os.Bundle
 
-import io.flutter.embedding.android.FlutterActivity
+import io.flutter.embedding.android.FlutterFragmentActivity
```
Comment (chadpav) [inline packages/health/example/android/app/src/main/kotlin/cachet/plugins/example_app/MainActivity.kt:5]: Add documentation for this step as well since the developer has to modify their Android application for this plugin to work now.

File: packages/health/example/android/app/src/main/AndroidManifest.xml [hunk 4]
```diff
@@ -79,6 +79,16 @@
                 <action android:name="androidx.health.ACTION_SHOW_PERMISSIONS_RATIONALE" />
             </intent-filter>
         </activity>
+        <activity-alias
```
Comment (chadpav) [inline packages/health/example/android/app/src/main/AndroidManifest.xml:82]: We should add something to the readme about this as part of the PR.

**Review by chadpav — APPROVED** [review]
Solution works and code LGTM. I think we should add some documentation to the readme though.

NOTE: I'm not a maintainer so my approval means nothing. :)

File: packages/health/example/lib/main.dart (hunk 1 above)
Comment (eliasteeny) [reply packages/health/example/lib/main.dart:163]: Yeah for some reason adding the sleep records is failing stating that it is an unsupported type. Although it is working on Android 13 and below.

File: packages/health/example/lib/util.dart (hunk 2 above)
Comment (eliasteeny) [reply packages/health/example/lib/util.dart:71]: Fetching the SLEEP data is crashing the app with the same error as writing.

**Comment by Letalus** [comment]
any updates on this PR? This is a critical bug for all android 34 devices

File: packages/health/example/lib/util.dart (hunk 2 above)
Comment (eric-nextsense) [reply packages/health/example/lib/util.dart:71]: I checked out this PR, uncommented the SLEEP lines, and i can read sleep data without a crash.

Comment (Letalus) [reply packages/health/example/lib/util.dart:71]: @eliasteeny could you please double check if this error still occurs on your device and if so, tell us your OS version and setup?

**Comment by hoffmatteo** [comment]
Hi, due to other projects and Christmas I was not able to properly evaluate it, I will do it this week though since it is an urgent issue.

File: packages/health/example/lib/util.dart (hunk 2 above)
Comment (eric-nextsense) [reply packages/health/example/lib/util.dart:71]: I got a closer look. I did not have a wearable so was reading only sleep session data (time asleep, one data point) and that worked fine. I get the crash now that i have a wearable and i try to read the stages. The issue seems to be that SleepStageRecord object no longer exists in Android 14. Instead there is a list in the SleepSessionRecord. You can see this sample code in https://developer.android.com/health-and-fitness/guides/health-connect/develop/sessions

suspend fun readSleepSessions(
    healthConnectClient: HealthConnectClient,
    startTime: Instant,
    endTime: Instant
) {
    val response =
        healthConnectClient.readRecords(
            ReadRecordsRequest(
                SleepSessionRecord::class,
                timeRangeFilter = TimeRangeFilter.between(startTime, endTime)
            )
        )
    for (sleepRecord in response.records) {
        // Retrieve relevant sleep stages from each sleep record
        val sleepStages = sleepRecord.stages
    }
}

The convertRecord function returns a single record and now it should return many, not sure how to handle this with the way it works now?

**Comment by Letalus** [comment]
@hoffmatteo I hope you had a great start in the new year. Just for an update, where you able to review @eliasteeny PR and have any comments on it?

File: packages/health/android/src/main/kotlin/cachet/plugins/health/HealthPlugin.kt [hunk 5]
```diff
@@ -1412,6 +1427,16 @@ class HealthPlugin(private var channel: MethodChannel? = null) :
         }
         binding.addActivityResultListener(this)
         activity = binding.activity
+
+
+        if ( healthConnectAvailable) {
+            val requestPermissionActivityContract = PermissionController.createRequestPermissionResultContract()
+
+            healthConnectRequestPermissionsLauncher =(activity as ComponentActivity).registerForActivityResult(requestPermissionActivityContract) { granted ->
```
Comment (madhav-sunil) [inline packages/health/android/src/main/kotlin/cachet/plugins/health/HealthPlugin.kt:1435]: Shouldn't we use `FlutterFragmentActivity` instead of `ComponentActivity`

**Comment by Alejandroem** [comment]
I have downloaded this fork and used the dependency as a reference from the download. And I can confirm this fix works. It will send me to the Health Connect Dialog to confirm the permissions and I can go from there.

I would advise anyone to do this until the package is properly updated. Thank you @eliasteeny you saved my app 💪🏻

**Comment by tomassasovsky** [comment]
Getting an error while testing on a Samsung Galaxy S23 Ultra, Android 14:

I/FLUTTER_HEALTH(27526): Permission launcher not found

**Comment by eric-nextsense** [comment]
I created a fork of this pull request (and created a pull request with it to update this one) which fixes sleep read issues.

https://github.com/eric-nextsense/flutter-plugins/tree/health-connect-android-14-fix

**Comment by tomassasovsky** [comment]
Works perfectly, thanks @eric-nextsense!

**Comment by limewaredev** [comment]
I'm getting the same error as @tomassasovsky - **Permission launcher not found**. I've tried referencing the https://github.com/eric-nextsense/flutter-plugins/tree/health-connect-android-14-fix repository in the pubspec.yaml file like this:

health:
    git:
      url: https://github.com/eric-nextsense/flutter-plugins
      path: packages/health
      ref: health-connect-android-14-fix

Any clues what might be wrong?

File: packages/health/android/src/main/kotlin/cachet/plugins/health/HealthPlugin.kt (hunk 5 above)
Comment (eliasteeny) [reply packages/health/android/src/main/kotlin/cachet/plugins/health/HealthPlugin.kt:1435]: We need the `ComponentActivity` in order to use the new api function `registerForActivityResult`, `FlutterFragmentActivity` is built upon the new Android apis which can be casted to a `ComponentActivity`. If you try to cast `FlutterActivity` to a `ComponentActivity` it will throw an error.

**Comment by eliasteeny** [comment]
> Getting an error while testing on a Samsung Galaxy S23 Ultra, Android 14:
> 
> I/FLUTTER_HEALTH(27526): Permission launcher not found



> I'm getting the same error as @tomassasovsky - **Permission launcher not found**. I've tried referencing the https://github.com/eric-nextsense/flutter-plugins/tree/health-connect-android-14-fix repository in the pubspec.yaml file like this:
> 
> ```
> health:
>     git:
>       url: https://github.com/eric-nextsense/flutter-plugins
>       path: packages/health
>       ref: health-connect-android-14-fix
> ```
> 
> Any clues what might be wrong?

Did you use `FlutterFragmentActivity` instead of `FlutterActivity` in the MainActivity file ?

**Comment by eric-nextsense** [comment]
Did you include the additional changes to fix sleep that are in my branch? They were not merged in this fork by the author.

https://github.com/eric-nextsense/flutter-plugins/tree/health-connect-android-14-fix

**Comment by eliasteeny** [comment]
@eric-nextsense I didn't merge it because I was waiting to have some free time to also fix the writing of sleep data.

**Comment by hoffmatteo** [comment]
> Did you include the additional changes to fix sleep that are in my branch? They were not merged in this fork by the author.
> 
> https://github.com/eric-nextsense/flutter-plugins/tree/health-connect-android-14-fix

Hi, I was just looking into the sleep issue. Your PR looks great, would it be possible for you to create the same PR targeting dev-health?

**Comment by tomo689** [comment]
> I'm getting the same error as @tomassasovsky - **Permission launcher not found**. I've tried referencing the https://github.com/eric-nextsense/flutter-plugins/tree/health-connect-android-14-fix repository in the pubspec.yaml file like this:
> 
> ```
> health:
>     git:
>       url: https://github.com/eric-nextsense/flutter-plugins
>       path: packages/health
>       ref: health-connect-android-14-fix
> ```
> 
> Any clues what might be wrong?

I also had the exact same problem and changed to use FlutterFragmentActivity instead of FlutterActivity. Now I got a new error message, "Access Denied (to Health Connect)!". Any clues to fix?
